### PR TITLE
Add token source based on sts token exchange

### PR DESCRIPTION
**Pull request by bianpengyuan**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: security/pkg/stsservice/tokenmanager/tokensource.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"golang.org/x/oauth2"
+
+	"istio.io/istio/security/pkg/stsservice"
+)
+
+// TokenSource specifies a oauth token source based on STS token exchange.
```
Comment (incfly) [inline security/pkg/stsservice/tokenmanager/tokensource.go:27]: an

Comment (incfly) [reply security/pkg/stsservice/tokenmanager/tokensource.go:27]: maybe mentioned the interface it implements?

File: security/pkg/stsservice/tokenmanager/tokensource.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+func (ts *TokenSource) setTokenManager(tm stsservice.TokenManager) {
+	ts.tm = tm
+}
+
+// Token returns Oauth token received from sts token exchange.
+func (ts *TokenSource) Token() (*oauth2.Token, error) {
+	params := stsservice.StsRequestParameters{
+		GrantType:        "urn:ietf:params:oauth:grant-type:token-exchange",
+		Scope:            ts.authScope,
```
Comment (incfly) [inline security/pkg/stsservice/tokenmanager/tokensource.go:53]: security/pkg/stsservice/server/server.go:38
constant var defined already

File: security/pkg/stsservice/tokenmanager/tokensource.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+type TokenSource struct {
+	tm           stsservice.TokenManager
+	subjectToken string
+	authScope    string
+}
+
+var _ oauth2.TokenSource = &TokenSource{}
+
+// NewTokenSource creates a token source based on STS token exchange.
+func NewTokenSource(trustDomain, subjectToken, authScope string) (*TokenSource, error) {
```
Comment (incfly) [inline security/pkg/stsservice/tokenmanager/tokensource.go:37]: no need to be error if it always succeed?

File: security/pkg/stsservice/tokenmanager/tokensource.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	return &TokenSource{
+		tm:           CreateTokenManager(GoogleTokenExchange, Config{TrustDomain: trustDomain}),
+		subjectToken: subjectToken,
+		authScope:    authScope,
+	}, nil
+}
+
+func (ts *TokenSource) setTokenManager(tm stsservice.TokenManager) {
+	ts.tm = tm
```
Comment (incfly) [inline security/pkg/stsservice/tokenmanager/tokensource.go:46]: no need for method if the test can access private method.

File: security/pkg/stsservice/tokenmanager/sts_test.go [hunk 5]
```diff
@@ -100,6 +100,39 @@ func TestStsCache(t *testing.T) {
...
+		t.Fatalf("failed to get access token: %v", err)
+	}
+	if got.AccessToken != mock.FakeAccessToken {
+		t.Errorf("access token want %v got %v", mock.FakeAccessToken, got.AccessToken)
+	}
+	if got.TokenType != "Bearer" {
+		t.Errorf("access token type want %v got %v", "Bearer", got.TokenType)
+	}
+	expireIn := time.Until(got.Expiry)
+	if expireIn < 3500*time.Second {
```
Comment (incfly) [inline security/pkg/stsservice/tokenmanager/sts_test.go:131]: @JimmyCYJ are we already having test like this? this is using a unexported implementation details in the mock token

File: security/pkg/stsservice/tokenmanager/tokensource.go (hunk 1 above)
Comment (bianpengyuan) [reply security/pkg/stsservice/tokenmanager/tokensource.go]: Done.

File: security/pkg/stsservice/tokenmanager/tokensource.go (hunk 2 above)
Comment (bianpengyuan) [reply security/pkg/stsservice/tokenmanager/tokensource.go:51]: Thanks! fixed.

File: security/pkg/stsservice/tokenmanager/tokensource.go (hunk 3 above)
Comment (bianpengyuan) [reply security/pkg/stsservice/tokenmanager/tokensource.go]: done.

File: security/pkg/stsservice/tokenmanager/tokensource.go (hunk 4 above)
Comment (bianpengyuan) [reply security/pkg/stsservice/tokenmanager/tokensource.go]: removed.

File: security/pkg/stsservice/tokenmanager/sts_test.go (hunk 5 above)
Comment (bianpengyuan) [reply security/pkg/stsservice/tokenmanager/sts_test.go]: Exposed expiry duration in the mock server package.

**Comment by bianpengyuan** [comment]
@incfly No, it will not be in oss.

**Comment by incfly** [comment]
/lgtm
will give @JimmyCYJ a chance to take look.

**Comment by bianpengyuan** [comment]
Ping @JimmyCYJ

**Pull request merged**